// mongodb-schema/lib/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.schemaStats = exports.SchemaAnalyzer = exports.getSimplifiedSchema = exports.getSchemaPaths = exports.parseSchema = void 0;
const schema_analyzer_1 = require("./schema-analyzer");
Object.defineProperty(exports, "SchemaAnalyzer", { enumerable: true, get: function () { return schema_analyzer_1.SchemaAnalyzer; } });
const schemaStats = __importStar(require("./stats"));
exports.schemaStats = schemaStats;
function verifyStreamSource(source) {
    if (!(Symbol.iterator in source) && !(Symbol.asyncIterator in source)) {
        throw new Error('Unknown input type for `docs`. Must be an array, ' +
            'stream or MongoDB Cursor.');
    }
    return source;
}
async function getCompletedSchemaAnalyzer(source, options) {
    const analyzer = new schema_analyzer_1.SchemaAnalyzer(options);
    for await (const doc of verifyStreamSource(source)) {
        analyzer.analyzeDoc(doc);
    }
    return analyzer;
}
async function parseSchema(source, options) {
    return (await getCompletedSchemaAnalyzer(source, options)).getResult();
}
exports.parseSchema = parseSchema;
async function getSchemaPaths(source) {
    return (await getCompletedSchemaAnalyzer(source)).getSchemaPaths();
}
exports.getSchemaPaths = getSchemaPaths;
async function getSimplifiedSchema(source) {
    return (await getCompletedSchemaAnalyzer(source)).getSimplifiedSchema();
}
exports.getSimplifiedSchema = getSimplifiedSchema;
exports.default = parseSchema;
